typescript/patterns/src/xa_transactions.ts

```diff
@@ -0,0 +1,123 @@
+import * as restate from "@restatedev/restate-sdk";
+import { randomUUID } from "crypto";
+
+//  -----------------                                         -----------------
+//                      Coordinating Database transactions
+//  -----------------                                         -----------------
+
+// Restate can be used to be the transaction coordinator for a database
+// transaction. The query to the database will be committed exactly once, and can
+// thus participate consistently in a sequence of steps in a workflow or durable
+// handler execution. 
+//
+// To do that, we use the 2-phase-commit protocol approach, implemented by many
+// databases via the XA transaction standard. For example, PostgreSQL supports this
+// via 'prepared transactions': https://www.postgresql.org/docs/current/sql-prepare-transaction.html
+//
+// Unlike the classical 2-phase-commit scenario, we are not using the protocol to
+// commit according to vote 
+
+// In principle, the operation is as simple as
+// (1) sideEffect: (execute query, prepare transaction under ID, return ID from sideEffect).
+//                 This side effect commits once and a single ID is stored in Restate for
+//                 committing.
+// (2) sideEffect: commit the prepared transaction
+//
+// The slightly higher complexity here comes from the fact that we need to ensure that
+// any transaction that was prepared, but will not be committed, must be cleaned up by
+// the recovery run. Otherwise that transaction will keep holding database locks and
+// block other transactions. We also re-try the transaction after that.
+
+// ----------------------------------------------------------------------------
+
+// The methods here must have the same semantics as in the PostgreSQL specification.
+// In pacticular the methods:
+//  - prepareTxn : https://www.postgresql.org/docs/current/sql-prepare-transaction.html
+//  - commitPreparedTxn : https://www.postgresql.org/docs/current/sql-commit-prepared.html
+//  - rollbackPreparedTxn : https://www.postgresql.org/docs/current/sql-rollback-prepared.html
+
+export interface Database<QueryT, ResultT> {
+
+    beginTxn(): Promise<void>    // initializes a new transaction
+    abortTxn(): Promise<void>    // aborts the current transaction
+
+    runSql(query: QueryT): Promise<ResultT>   // executed a sql statement
+
+    prepareTxn(txnId: string): Promise<void>          // prepared txn for commit under given ID
+    commitPreparedTxn(txnId: string): Promise<void>   // commits a txn prepared under given ID
+    rollbackPreparedTxn(txnId: string): Promise<void> // rolls back a txn prepared under given ID
+}
+
+
+export async function runXaDatabaseTransaction<QueryT, ResultT>(
```

Comment:
I think this example breaks in multiple places:
  - through the durable execution, we always follow the path where the txn ID is initially unset.
  - A failure between `await database.prepareTxn(txnId);` and `await database.commitPreparedTxn(txnId);` means the prepared transaction stays. Because it won't ever be cleaned up (point 1), it will block the database forever.
  - Imagine you have a process that makes it to `const result = await database.runSql(query);` and then stalls. A second (failover execution) makes it all the way though, then the stalled process continues. You now have committed everything twice.
  
Those are exactly the points I was trying to address in my implementation. Maybe there is a slightly simpler way than I did it, but I think this is inherently complex, because
  - (a) Postgres misses some primitives to make this easier (like fence off previous IDs)
  - (b) we are implementing the txn coordinator, with all recovery logic
  - (c) for that txn coordinator, durable execution is surprisingly not the perfect match, because we want to look at the log while replaying, not skip over actions just like that.

Reply:
I think the hardest problems (and the main class of errors Jepsen tends to find) are related to
  - partial network partitions
  - long stalls in processes that make another process think it timed out and disappeared (but where the process then comes back). Rare corner cases, but saw those problems actually happening during container migrations.